pkg/model/openstackmodel/servergroup.go

```diff
@@ -308,57 +313,75 @@ func (b *ServerGroupModelBuilder) Build(c *fi.CloudupModelBuilderContext) error
 			b.associateFIPToKeypair(lbfipTask)
 		}
 
-		poolTask := &openstacktasks.LBPool{
-			Name:         fi.PtrTo(fmt.Sprintf("%s-https", fi.ValueOf(lbTask.Name))),
-			Loadbalancer: lbTask,
-			Lifecycle:    b.Lifecycle,
+		portsToListen := map[string]int{"https": wellknownports.KubeAPIServer}
+		if b.Cluster.UsesNoneDNS() {
+			portsToListen["kops-controller"] = wellknownports.KopsControllerPort
 		}
-		c.AddTask(poolTask)
 
-		listenerTask := &openstacktasks.LBListener{
-			Name:      lbTask.Name,
-			Lifecycle: b.Lifecycle,
-			Pool:      poolTask,
-		}
-		if useVIPACL {
-			var AllowedCIDRs []string
-			// currently kOps openstack supports only ipv4 addresses
-			for _, CIDR := range b.Cluster.Spec.API.Access {
-				if net.IsIPv4CIDRString(CIDR) {
-					AllowedCIDRs = append(AllowedCIDRs, CIDR)
-				}
+		for name, port := range portsToListen {
+			poolTask := &openstacktasks.LBPool{
+				Name:         fi.PtrTo(fmt.Sprintf("%s-%s", fi.ValueOf(lbTask.Name), name)),
+				Loadbalancer: lbTask,
+				Lifecycle:    b.Lifecycle,
 			}
-			sort.Strings(AllowedCIDRs)
-			listenerTask.AllowedCIDRs = AllowedCIDRs
-		}
-		c.AddTask(listenerTask)
+			c.AddTask(poolTask)
 
-		monitorTask := &openstacktasks.PoolMonitor{
-			Name:      lbTask.Name,
-			Pool:      poolTask,
-			Lifecycle: b.Lifecycle,
-		}
-		c.AddTask(monitorTask)
+			nameForResource := fi.ValueOf(lbTask.Name)
+			if isKopsController(name) {
+				nameForResource = fi.ValueOf(lbTask.Name) + "-kops-controller"
+			}
 
-		ifName, err := b.GetNetworkName()
-		if err != nil {
-			return err
-		}
-		for _, mastersg := range masters {
-			associateTask := &openstacktasks.PoolAssociation{
-				Name:          mastersg.Name,
-				Pool:          poolTask,
-				ServerGroup:   mastersg,
-				InterfaceName: fi.PtrTo(ifName),
-				ProtocolPort:  fi.PtrTo(443),
-				Lifecycle:     b.Lifecycle,
-				Weight:        fi.PtrTo(1),
+			listenerTask := &openstacktasks.LBListener{
+				Name:      fi.PtrTo(nameForResource),
+				Port:      fi.PtrTo(port),
+				Lifecycle: b.Lifecycle,
+				Pool:      poolTask,
 			}
 
-			c.AddTask(associateTask)
+			if useVIPACL {
+				var AllowedCIDRs []string
+				// currently kOps openstack supports only ipv4 addresses
+				for _, CIDR := range b.Cluster.Spec.API.Access {
+					if net.IsIPv4CIDRString(CIDR) {
+						AllowedCIDRs = append(AllowedCIDRs, CIDR)
+					}
+				}
+				sort.Strings(AllowedCIDRs)
+				if isKopsController(name) {
```

Comment:
I think other cloudproviders is allowing kops controller port to same targets than API itself. But I do not see any reason why that should be allowed from outside cluster?

Reply:
I did that to not allow `0.0.0.0/0` for public topologies.
In particular, GCP is quite inflexible when using a public LB, as it cannot allow communication with the private net, only public IPs, which are unknown. So, you are right, it's a hack that will be fixed at some point.
If you can, allow access to that port only from the cluster network CIDR.

Reply:
Though, the kops-controller access is required when using it for bootstrap.
See: https://github.com/kubernetes/kops/pull/14460.

Reply:
Openstack does not use kops controller for bootstrap

Reply:
I guess nodes don't need access to kops-controller. It would be nice to use kops-controller for bootstrap though.

Reply:
what is the advantage of using kops-controller for bootstrapping? We do not need that much information in metadata in future in normal nodes?

Reply:
You do need config and cert(s) passed to nodes. This requires tokens to read the config and cert(s) files from s3 or similar.